### ISSUE TITLE
Cannot find module '@vercel/build-utils' or its corresponding type declarations.

### ISSUE THREAD

**Issue by leerob** [zeit/now#7951]
### System Info

- yarn: `1.22.17`
- pnpm: `7.1.6`
- npm: `8.1.2`

All three had the same issue.

### Dependencies

- `"@types/node": "^17.0.42"`
- `"@vercel/node": "^2.1.0"`
- `"typescript": "^4.7.3"`

### Error

```
node_modules/.pnpm/@vercel+node@2.1.0/node_modules/@vercel/node/dist/index.d.ts:1:29 - error TS2307: Cannot find module '@vercel/build-utils' or its corresponding type declarations.

1 import { shouldServe } from '@vercel/build-utils';

node_modules/.pnpm/@vercel+node@2.1.0/node_modules/@vercel/node/dist/index.d.ts:2:60 - error TS2307: Cannot find module '@vercel/build-utils' or its corresponding type declarations.

2 import type { BuildV3, PrepareCache, StartDevServer } from '@vercel/build-utils';
```

## Reproduction

```js
import type { VercelRequest, VercelResponse } from '@vercel/node';
import type { Readable } from 'node:stream';

export const config = {
  api: {
    bodyParser: false,
  },
};

async function buffer(readable: Readable) {
  const chunks = [];
  for await (const chunk of readable) {
    chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : chunk);
  }
  return Buffer.concat(chunks);
}

export default async function (req: VercelRequest, res: VercelResponse) {
  if (req.method === 'POST') {
    const buf = await buffer(req);
    const rawBody = buf.toString('utf8');

    res.json({ rawBody });
  } else {
    res.setHeader('Allow', 'POST');
    res.status(405).end('Method Not Allowed');
  }
}
```

**Comment by MichaelDeBoey** [comment]
This should be fixed by #7934 I guess?

**Comment by MichaelDeBoey** [comment]
@styfle After using the updated version in https://github.com/remix-run/remix/pull/3420, it seems like `@vercel/build-utils` is missing `@vercel/frameworks` & `@vercel/routing-utils` in it's `dependencies` list.
https://github.com/remix-run/remix/runs/6924983939?check_suite_focus=true#step:6:124

```sh
Error: node_modules/@vercel/build-utils/dist/detect-builders.d.ts(1,23): error TS2307: Cannot find module '@vercel/routing-utils' or its corresponding type declarations.
Error: node_modules/@vercel/build-utils/dist/detect-framework.d.ts(1,27): error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.
Error: node_modules/@vercel/build-utils/dist/workspaces/workspace-managers.d.ts(1,32): error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.
Error: node_modules/@vercel/build-utils/dist/monorepos/monorepo-managers.d.ts(1,32): error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.
```

I however think it would be better to remove `@vercel/build-utils` from `@vercel/node` instead tbh.

**Comment by styfle** [comment]
I was afraid that might happen.

We probably need to separate `@vercel/build-utils` into multiple packages because none of those functions are used by `@vercel/node` or other builders, so they probably don't belong in `@vercel/build-utils`.

@TooTallNate @EndangeredMassa Putting this on your radar since we briefly discussed this once before.

**Comment by MichaelDeBoey** [comment]
@styfle `@vercel/node` is indeed only using the following function from `@vercel/build-utils` (which all don't rely on `@vercel/frameworks` & `@vercel/routing-utils`):

- `debug`
- `download`
- `EdgeFunction`
- `FileBlob`
- `FileFsRef`
- `getNodeVersion`
- `getSpawnOptions`
- `glob`
- `isSymbolicLink`
- `NodejsLambda`
- `NowBuildError`
- `runNpmInstall`
- `runPackageJsonScript`
- `shouldServe`
- `streamToBuffer`
- `walkParentDirs`

And the following types:
- `BuildResultV3`
- `BuildV3`
- `Config`
- `File`
- `Files`
- `Meta`
- `NodeVersion`
- `PrepareCache`
- `StartDevServer`
- `StartDevServerOptions`

**Comment by MichaelDeBoey** [comment]
@EndangeredMassa @styfle @TooTallNate Is there anything I can do to help get this fixed?
I'm happy to create a PR if you can point me into the direction you want to take this.

**Comment by styfle** [comment]
Its more than just `@vercel/node` using `@vercel/build-utils`, its used by every builder and its injected automatically, so we can't easily remove functions without causing deployments to fail.

I think step one is to figure out when this regression was introduced (it must have been working at some point).

Then we can figure out how to split it up `@vercel/build-utils` from there, perhaps its just one function that was added that needs to be moved into a separate package.

**Comment by MichaelDeBoey** [comment]
The problems began when using v2 of `@vercel/node`, so I think it would be a good starting point to look at what changed between v1 & v2 then.

**Comment by styfle** [comment]
Here are the changes to `@vercel/node` between 1.15.4 and 2.0.0: https://github.com/vercel/vercel/compare/%40vercel/node%401.15.4...%40vercel/node%402.0.0

However, I think the issue is actually in `@vercel/build-utils`, so we'll take some time to split up `@vercel/build-utils` into separate packages on Tuesday or Wednesday.

**Comment by davidhartsough** [comment]
I've been having the same issues, so I've tried going backwards one version at a time between the packages `@vercel/node` and `vercel` until the errors went away. I assumed it was an issue with `@vercel/node` from this thread, so I started there and went back to the last version I was using before when my project was working without errors. However, I went all the way back to that point and noticed the errors were still there. So then I left it and started going back in `vercel` versions. 

I actually discovered that the error begins for me only with the `vercel` package, and it has nothing to do with the `@vercel/node` package. I can have `@vercel/node@2.3.1` (latest) with no issues, but as soon as I bump `vercel` from `vercel@24.2.4` to `24.2.5`, I get the errors. 

For version `24.2.4`, the dependencies are: 

```shell
├─ @vercel/build-utils@3.1.0
├─ @vercel/go@1.4.3
├─ @vercel/python@2.3.3
├─ @vercel/ruby@1.3.6
└─ vercel@24.2.4
```

For version `24.2.5`, the deps are: 


<details>

 <summary>(Dependencies)</summary>

```shell
├─ @babel/runtime@7.18.6
├─ @mapbox/node-pre-gyp@1.0.9
├─ @remix-run/node@1.4.3
├─ @remix-run/server-runtime@1.4.3
├─ @remix-run/vercel@1.4.3
├─ @types/busboy@0.3.2
├─ @types/cookie@0.4.1
├─ @types/node-fetch@2.6.2
├─ @vercel/build-utils@3.1.1
├─ @vercel/go@1.4.4
├─ @vercel/next@2.9.0
├─ @vercel/python@2.3.4
├─ @vercel/redwood@0.8.4
├─ @vercel/remix@0.0.2
├─ @vercel/routing-utils@1.13.3
├─ @vercel/ruby@1.3.7
├─ @vercel/static-build@0.26.0
├─ @web-std/blob@3.0.4
├─ @web-std/file@3.0.2
├─ @web-std/stream@1.0.0
├─ @zxing/text-encoding@0.9.0
├─ aproba@1.2.0
├─ are-we-there-yet@2.0.0
├─ bindings@1.5.0
├─ blob-stream@0.1.3
├─ blob@0.0.4
├─ busboy@0.3.1
├─ chownr@1.1.4
├─ code-point-at@1.1.0
├─ color-support@1.1.3
├─ cookie-signature@1.2.0
├─ cookie@0.4.2
├─ core-util-is@1.0.3
├─ detect-libc@1.0.3
├─ dicer@0.3.0
├─ estree-walker@2.0.2
├─ file-uri-to-path@1.0.0
├─ fs-minipass@1.2.7
├─ function.prototype.name@1.1.5
├─ gauge@3.0.2
├─ has-unicode@2.0.1
├─ ignore-walk@3.0.4
├─ is-arguments@1.1.1
├─ is-generator-function@1.0.10
├─ is-typed-array@1.1.9
├─ isarray@1.0.0
├─ jsesc@3.0.2
├─ minizlib@1.3.3
├─ mkdirp@0.5.6
├─ needle@2.9.1
├─ node-gyp-build@4.4.0
├─ node-pre-gyp@0.13.0
├─ nopt@4.0.3
├─ npm-bundled@1.1.2
├─ npm-packlist@1.4.8
├─ npmlog@4.1.2
├─ number-is-nan@1.0.1
├─ object-assign@4.1.1
├─ os-homedir@1.0.2
├─ osenv@0.1.5
├─ path-to-regexp@6.1.0
├─ process-nextick-args@2.0.1
├─ react-router-dom@6.3.0
├─ react-router@6.3.0
├─ regenerator-runtime@0.13.9
├─ regexp.prototype.flags@1.4.3
├─ rollup-pluginutils@2.8.2
├─ sax@1.2.4
├─ set-blocking@2.0.0
├─ set-cookie-parser@2.5.0
├─ streamsearch@0.1.2
├─ tar@4.4.19
├─ util@0.12.4
├─ vercel@24.2.5
├─ web-encoding@1.1.5
├─ web-streams-polyfill@3.2.1
├─ which-typed-array@1.1.8
└─ wide-align@1.1.5
```

</details>

The only change in `@vercel/build-utils` is bumping from `v3.1.0` to `@3.1.1`.

The errors I get are: 

<details>

 <summary>(Errors)</summary>

```shell
node_modules/@vercel/node/node_modules/@vercel/build-utils/dist/detect-framework.d.ts:1:27 - error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.

1 import { Framework } from '@vercel/frameworks';
                            ~~~~~~~~~~~~~~~~~~~~

node_modules/@vercel/node/node_modules/@vercel/build-utils/dist/fs/glob.d.ts:1:26 - error TS7016: Could not find a declaration file for module 'glob'. '/Users/davidhartsough/dev/thought-saver/ts-github-repos/thought-saver-api/node_modules/glob/glob.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/glob` if it exists or add a new declaration (.d.ts) file containing `declare module 'glob';`

1 import vanillaGlob_ from 'glob';
                           ~~~~~~

node_modules/@vercel/node/node_modules/@vercel/build-utils/dist/monorepos/monorepo-managers.d.ts:1:32 - error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.

1 import type { Framework } from '@vercel/frameworks';
                                 ~~~~~~~~~~~~~~~~~~~~

node_modules/@vercel/node/node_modules/@vercel/build-utils/dist/workspaces/workspace-managers.d.ts:1:32 - error TS2307: Cannot find module '@vercel/frameworks' or its corresponding type declarations.

1 import type { Framework } from '@vercel/frameworks';
                                 ~~~~~~~~~~~~~~~~~~~~


Found 4 errors.
```

</details>


So @MichaelDeBoey [might be right (comment above)](https://github.com/vercel/vercel/issues/7951#issuecomment-1158069189) 

> it seems like `@vercel/build-utils` is missing `@vercel/frameworks` & `@vercel/routing-utils` in it's `dependencies` list. 

And although I know little to nothing about these details, I think I would support this idea: 

> I however think it would be better to remove @vercel/build-utils from @vercel/node instead tbh.

Because the project I have doesn't have a build step, a framework, etc.

**Comment by styfle** [comment]
I created PR #8054

**Comment by styfle** [comment]
I did some more digging and found the root cause. This PR will fix it #8064
